Source/MOLXPCConnection/MOLXPCConnection.m

```diff
@@ -150,7 +154,33 @@ - (void)resume {
   }
 }
 
+- (void)allowUnprivilegedClients:(BOOL)enable {
+  _allowUnprivilegedClients = enable;
+}
+
 - (BOOL)listener:(NSXPCListener *)listener shouldAcceptNewConnection:(NSXPCConnection *)connection {
+  // Fail this connection if it's from an unprivileged user and we have been
+  // configured to only allow root/admins
+  BOOL userIsPrivileged = [connection effectiveUserIdentifier] == 0;
+  if (!userIsPrivileged && !_allowUnprivilegedClients) {
+    return NO;
+  }
+
+  // Get the privileged interface for privileged users, and fallback to unprivileged
+  // if the first interface is not available; do the opposite for unprivileged users
+  NSXPCInterface *preferred_interface;
+  NSXPCInterface *fallback_interface;
+
+  if (userIsPrivileged) {
+    preferred_interface = self.privilegedExportedInterface;
+    fallback_interface = self.unprivilegedExportedInterface;
+  } else {
+    preferred_interface = self.unprivilegedExportedInterface;
+    fallback_interface = self.privilegedExportedInterface;
+  }
+
+  NSXPCInterface *userInterface = (preferred_interface ? preferred_interface : fallback_interface);
```

Comment:
I think this could be simpler and more explicit about which interface to export; if we remove the `allowUnprivilegedClients` property as I suggested above and require that the right interface is used for privileged/unprivileged clients, this can become:

```objc
NSXPCInterface *interface;
if ([connection effectiveUserIdentifier] == 0) {
  interface = self.privilegedInterface;
} else {
  interface = self.unprivilegedInterface;
}

// TODO(any): Remove 1-2 releases after exportedInterface was marked deprecated.
if (!interface) {
  interface = self.exportedInterface;
}

if (!interface) return NO;

.....

connection.exportedInterface = interface;
```

Clients which then want the same interface exported for both privileged and unprivileged clients can use:
```conn.privilegedExportedInterface = conn.unprivilegedExportedInterface = myNewInterface;```

And if the appropriate interface is unset (and assuming exportedInterface isn't being used) the connection will be rejected.